src/utils.ts

```diff
@@ -54,19 +57,59 @@ export type ProcessWithOrderingParams<T> = {
  *
  * This same scenario is true for SQS FIFO queues, which will order messages
  * by MessageGroupId.
+ *
+ * When rejectOnError is false, the method will return the list
+ * of unprocessed items by listId. Callers will then have access to the error
+ * that caused the list to stop processing events, plus all the events
+ * that were not processed in the batch.
  */
 export const processWithOrdering = async <T>(
   params: ProcessWithOrderingParams<T>,
   process: (item: T) => Promise<void>,
 ) => {
-  const lists = Object.values(groupBy(params.items, params.orderBy));
+  const rejectOnError = params.rejectOnError ?? true;
+  const groupedItems = groupBy(params.items, params.orderBy);
+  const listIds = Object.keys(groupedItems);
+  const lists = Object.values(groupedItems);
+  const unprocessedRecordsByListId = zipObject<{ error: any; items: T[] }>(
+    listIds,
+    lists.map(() => ({ error: null, items: [] })),
+  );
+
   await pMap(
     lists,
-    async (list) => {
-      for (const item of list) {
-        await process(item);
+    async (list, listIndex) => {
+      for (let i = 0; i < list.length; i++) {
+        const item = list[i];
+
+        try {
+          await process(item);
+        } catch (error) {
+          // Keep track of all unprocessed items and stop processing the current
+          // list as soon as we encounter the first error.
+          unprocessedRecordsByListId[listIds[listIndex]] = {
+            error,
+            items: list.slice(i),
+          };
+          return;
+        }
       }
     },
-    { concurrency: params.concurrency, stopOnError: params.stopOnError },
+    {
+      concurrency: params.concurrency,
+    },
   );
+
+  const aggregateErrors = Object.values(unprocessedRecordsByListId)
+    .map((record) => record.error)
+    .flat();
+
+  if (aggregateErrors.length > 0 && rejectOnError) {
+    throw new AggregateError(aggregateErrors);
+  }
+
+  return {
+    hasUnprocessedRecords: aggregateErrors.length > 0,
+    unprocessedRecords: unprocessedRecordsByListId,
+  };
```

Comment:
I think that would have been ok if this helper was only used for handling SQS. But it's also used for DynamoDB streams, where there is no notion of partial batches. 

In those cases, I think we should throw directly.

My thought when making changes to this API was to avoid having to change the DynamoDB calling site. If we don't throw, we'd have to make that place check for unprocessed items and throw itself. To me that seems like a leaky abstraction for any DynamoDB user. 

With the current API we get a nice default (`rejectOnError` is `true` by default), and a nice escape hatch for SQS by just setting `rejectOnError` to `false`.

Reply:
I didn't state it clearly in the original comment -- yes, we'd definitely need to update the DynamoDB call site.

I'm not sure I'm following about the "leaky abstraction" -- the details of all of this implementation are hidden from people actually using the DynamoDB / SQS abstractions. What abstraction is leaking?

As-is, I'm finding it a bit hard to reason about the behavior of `processWithOrdering`, since it can behave really differently depending on it's params (sometimes throws on failures, sometimes doesn't).

The simplification makes the behavior of this function _really_ explicit:
- it processes records in a particular order
- it never throws
- it returns records that weren't successfully processed

I think the win in simplicity + clarity is worth updating the DynamoDB call site to explicitly throw.